src/Request/SoapPhpRequest.php

```diff
@@ -0,0 +1,262 @@
+<?php
+
+declare(strict_types=1);
+
+namespace RicorocksDigitalAgency\Soap\Request;
+
+use RicorocksDigitalAgency\Soap\Contracts\Builder;
+use RicorocksDigitalAgency\Soap\Contracts\PhpSoap\Client;
+use RicorocksDigitalAgency\Soap\Contracts\Request;
+use RicorocksDigitalAgency\Soap\Contracts\Soapable;
+use RicorocksDigitalAgency\Soap\Response\Response;
+use RicorocksDigitalAgency\Soap\Support\Header;
+use RicorocksDigitalAgency\Soap\Support\Tracing\Trace;
+use SoapHeader;
+
+/**
+ * @internal
+ */
+final class SoapPhpRequest implements Request
+{
+    private Builder $builder;
+
+    private Client $client;
+
+    private Response $response;
+
+    private string $endpoint;
+
+    private string $method;
+
+    /**
+     * @var array<string, mixed>|Soapable
+     */
+    private $body = [];
+
+    /**
+     * @var array{beforeRequesting: array<int, callable(Request): mixed>, afterRequesting: array<int, callable(Request, Response): mixed>}
+     */
+    private array $hooks = [
+        'beforeRequesting' => [],
+        'afterRequesting' => [],
+    ];
+
+    /**
+     * @var array<string, mixed>
+     */
+    private array $options = [];
+
+    /**
+     * @var array<int, Header>
+     */
+    private array $headers = [];
+
+    public function __construct(Builder $builder, Client $client)
+    {
+        $this->builder = $builder;
+        $this->client = $client;
+    }
+
+    public function to(string $endpoint): self
+    {
+        $this->endpoint = $endpoint;
+
+        return $this;
+    }
+
+    /**
+     * @param array<mixed> $parameters
+     */
+    public function __call(string $name, array $parameters): mixed
+    {
+        return $this->call($name, $parameters[0] ?? []);
+    }
+
+    /**
+     * @param array<string, mixed>|Soapable $body
+     */
+    public function call(string $method, array|Soapable $body = []): Response
+    {
+        $this->method = $method;
+        $this->body = $body;
+
+        collect($this->hooks['beforeRequesting'])->each(fn ($callback) => $callback($this));
+        $this->body = $this->builder->handle($this->body);
+
+        $response = $this->getResponse();
+        collect($this->hooks['afterRequesting'])->each(fn ($callback) => $callback($this, $response));
+
+        return $response;
+    }
+
+    private function getResponse(): Response
+    {
+        return $this->response ??= $this->getRealResponse();
+    }
+
+    private function getRealResponse(): Response
+    {
+        return tap(
+            Response::new($this->makeRequest()),
+            fn ($response) => data_get($this->options, 'trace')
+                ? $response->setTrace(Trace::client($this->client()))
+                : $response
+        );
+    }
+
+    private function makeRequest(): mixed
+    {
+        return $this->client()->call($this->getMethod(), $this->getBody());
+    }
+
+    private function client(): Client
+    {
+        return $this->client
```

Comment:
This structure is a bit strange since you can change options in between requests, but the decorated client itself uses a memoized versionof the client. Only the headers will be changed since the options and endpoint might have been set.

Either way : you want to avoid constructing the internal SoapClient twice, since it will load the wsdl again as well. which canbe a slow task